feat: implementation of streaming mode

blofroth opened the pull request:
based on InputStreams

```
# producer
# Need to have KafkaProducer<T, InputStream>
# Need to send header message-claim-check-payload-size as Long
interceptor.classes=se.irori.kafka.claimcheck.ClaimCheckStreamingProducerInterceptor
value.serializer.wrapped.serializer=se.irori.kafka.claimcheck.InputStreamSerializer

# consumer
# Need to have KafkaConsumer<T, InputStream>
value.deserializer=se.irori.kafka.claimcheck.ClaimCheckStreamingDeserializer
```

blofroth commented:
Perhaps we should instead allow users of the library to specify the streaming message size header:
`claimcheck.message.size.long.header`

Then we could allow out of the box integration with e.g. Camel Kafka, Camel SFTP etc

CamelFileLength (consumer) Constant: FILE_LENGTH | A long value containing the file size. |   | long
https://camel.apache.org/components/3.17.x/file-component.html#_message_headers

norbacker commented on claim-check-interceptors-azure/src/main/java/se/irori/kafka/claimcheck/azure/AzureBlobStorageClaimCheckBackend.java:
```diff
@@ -47,6 +48,24 @@ public ClaimCheck checkIn(ProducerRecord<byte[], byte[]> largeRecord) {
   }
 
+  @Override
+  public ClaimCheck checkInStreaming(String topic, InputStream payload, long payloadSize) {
```
This method and `checkIn` should share the similar code in some way

norbacker commented on claim-check-interceptors-azure-8/src/main/java/se/irori/kafka/claimcheck/azurev8/AzureBlobStorageClaimCheckBackendV8.java:
```diff
@@ -64,6 +65,11 @@ public ClaimCheck checkIn(ProducerRecord<byte[], byte[]> largeRecord) {
 
+  @Override
+  public ClaimCheck checkInStreaming(String topic, InputStream payload, long payloadSize) {
+    throw new UnsupportedOperationException("Not implemented for Azure V8 backend");
```
Is it much work to do the v8 implementation as well?

blofroth replied:
I haven't looked, but I think eventually that whole implementation should be removed since v8 SDK isn't kept very up to date. But perhaps we should have some alternative implementation for people who are getting dependency collisions, perhaps
https://jclouds.apache.org/

blofroth replied:
Turned out to be easy to implement so I just did it for v8

norbacker commented on claim-check-interceptors-azure/src/main/java/se/irori/kafka/claimcheck/azure/AzureBlobStorageClaimCheckBackend.java:
```diff
@@ -69,6 +88,25 @@ public byte[] checkOut(ClaimCheck claimCheck) {
   }
 
+  @Override
+  public InputStream checkOutStreaming(ClaimCheck claimCheck) {
```
This method and `checkOut` should share the similar code in some way

norbacker commented on claim-check-interceptors-core/src/main/java/se/irori/kafka/claimcheck/ClaimCheckBackend.java:
```diff
@@ -20,6 +22,8 @@ public interface ClaimCheckBackend extends Configurable  {
    */
   ClaimCheck checkIn(ProducerRecord<byte[], byte[]> largeRecord);
 
+  ClaimCheck checkInStreaming(String topic, InputStream payload, long payloadSize);
```
Docs missing

norbacker commented on claim-check-interceptors-core/src/main/java/se/irori/kafka/claimcheck/ClaimCheckBackend.java:
```diff
@@ -28,6 +32,8 @@ public interface ClaimCheckBackend extends Configurable  {
    */
   byte[] checkOut(ClaimCheck claimCheck);
 
+  InputStream checkOutStreaming(ClaimCheck claimCheck);
```
Docs missing

norbacker reviewed (approved):
Seems like a good feature. But it's very rough. Would it be possible to implement this in a fork and do a proper implementation after the vacation?

blofroth commented on claim-check-interceptors-core/src/main/java/se/irori/kafka/claimcheck/ClaimCheckStreamingProducerInterceptor.java:
```diff
@@ -0,0 +1,170 @@
+      byte[] longBytes =
+          producerRecord.headers().lastHeader(HEADER_MESSAGE_CLAIM_CHECK_PAYLOAD_SIZE).value();
+      ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+      buffer.put(longBytes);
```
Hmm, that is probably just me messing up long serialization. It is not as if it is a big buffer. I probably just need to test it with some unit tests.

blofroth commented:
I have addressed most comments, added unit and integration tests. So it should be much less rough now. 

Possibly I could try adding a dynamic size header, or we leave that for later.

blofroth commented:
I created a Quarkus/camel example project to test it out. When we are happy with it and the feature we can make it public
https://github.com/irori-ab/quarkus-kafka-claim-check-example

It seems to work from initial test. It uses less memory then the total file size at least. Maybe we can test it with some restricted heap sizes to sanity check

luciandragos reviewed (approved):
Looks good to me, we have been using this feature branch for our client for a while and didn't face any additional issues.
